src/leap/keymanager/openpgp.py

```diff
@@ -447,23 +434,30 @@ def _put_key_doc(self, key):
         def check_and_put(docs, key):
             if len(docs) == 1:
                 doc = docs.pop()
-                if key.fingerprint == doc.content[KEY_FINGERPRINT_KEY]:
+                oldkey = build_key_from_dict(OpenPGPKey, doc.content)
+                if key.fingerprint == oldkey.fingerprint:
                     # in case of an update of the key merge them with gnupg
                     with self._temporary_gpgwrapper() as gpg:
-                        gpg.import_keys(doc.content[KEY_DATA_KEY])
+                        gpg.import_keys(oldkey.key_data)
                         gpg.import_keys(key.key_data)
                         gpgkey = gpg.list_keys(secret=key.private).pop()
-                        key = _build_key_from_gpg(
+                        mergedkey = self._build_key_from_gpg(
                             gpgkey,
                             gpg.export_keys(gpgkey['fingerprint'],
                                             secret=key.private))
-                    doc.set_json(key.get_json())
+                    mergedkey.validation = max(
```

Comment:
If an attacker manages to push you the same key, but with lower validation level that should not allow her to downgrade your validation level for this key.

One example:
1. Alice checks the fingerprint of Bob, and puts the key in her keymanager with validation level _fingerprint_.
2. Eve sends an email to Alice impersonating Bob with the Bob public key as attachment.
3. Alice gets Bob's key from the fake email of Eve and downgrade the validation level to _weak chain_.
4. Bob's key expires and Eve impersonates Bob again sending an email to Alice with a new public key that she has created, before Bob has time to update his expiration date.
5. Alice accepts this new public key as it has the same validation level and now communicates with Bob encrypting everything with Eve's key.